### PR TITLE
Fixed validation stats and added new summary stats

### PR DESCRIPTION
**Pull request by haim-barad**
The changes are in the Early Exit path only. Tested for training, validation and evaluate phases.

### CODE REVIEW

File: examples/classifier_compression/compress_classifier.py [hunk 1]
```diff
@@ -645,28 +652,35 @@ def earlyexit_loss(output, target, criterion, args):
...
+                args.exiterrors[exitnum].add(torch.tensor(np.array(output[exitnum].data[batch_index], ndmin=2)),
+                        torch.full([1], target[batch_index], dtype=torch.long))
                 args.exit_taken[exitnum] += 1
-            else:
-                # skip the early exits and include results from end of net
-                args.exiterrors[args.num_exits-1].add(torch.tensor(np.array(output[args.num_exits-1].data[batchnum],
-                                                                            ndmin=2)),
-                        torch.full([1], target[batchnum], dtype=torch.long))
-                args.exit_taken[args.num_exits-1] += 1
+                earlyexit_taken = True
```
Comment (nzmora) [inline examples/classifier_compression/compress_classifier.py]: This is part of an ever-growing else-clause.  I think this else-clause is large enough to merit refactoring it to a separate function.  What say you?

File: examples/classifier_compression/compress_classifier.py [hunk 2]
```diff
@@ -570,8 +570,9 @@ def _validate(data_loader, model, criterion, loggers, args, epoch=-1):
                 if args.display_confusion:
                     confusion.add(output.data, target)
             else:
-                # If using Early Exit, then compute outputs at all exits - output is now a list of all exits
-                # from exit0 through exitN (i.e. [exit0, exit1, ... exitN])
+                # If using Early Exit, then compute outputs at all exits
+                # output is now a list of all exit probabilities from
+                # exit0 through exitN (i.e. [exit0, exit1, ... exitN])
```
[the remark this answers is not in the document]
Comment (haim-barad) [reply examples/classifier_compression/compress_classifier.py]: Agreed - I'll merge them into the comments already heading the earlyexit_validate_loss() function.

Better yet, it might make more sense to describe what's happening outside the code in the earlyexit.md file that's in the documentation section (we need a link to it from somewhere).

Comments?

File: examples/classifier_compression/compress_classifier.py [hunk 3]
```diff
@@ -627,8 +628,14 @@ def _validate(data_loader, model, criterion, loggers, args, epoch=-1):
             if args.exit_taken[exitnum]:
```
[the remark this answers is not in the document]
Comment (haim-barad) [reply examples/classifier_compression/compress_classifier.py]: Yes and no. I thought about doing that, but I wanted to print out the % of data going out each exit before the summary stats about the accuracy of the exits. So you're suggesting that it would make more sense to just print out everything about exit0, then everything about exit1, and so on... I'm not sure if the comparison would be as clear (even though it would save a loop).

File: examples/classifier_compression/compress_classifier.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (netaz) [reply examples/classifier_compression/compress_classifier.py]: Yes, we should add the earlyexit.md file to [here](https://nervanasystems.github.io/distiller/) and we can also link to it directly from the [README](https://github.com/NervanaSystems/distiller/blob/master/README.md).  I've recently added a mention of it (search for "early exit").  Let's review the content of earlyexit.md.

File: examples/classifier_compression/compress_classifier.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (netaz) [reply examples/classifier_compression/compress_classifier.py]: OK, I see... if you think it makes the user-experience better, then you should definitely keep the two loops.

File: examples/classifier_compression/compress_classifier.py (hunk 1 above)
Comment (netaz) [reply examples/classifier_compression/compress_classifier.py]: No answer means you're not happy? ;-)
